src/Utilities/vector_bounds.jl

```diff
@@ -0,0 +1,411 @@
+    S1 = _flag_to_set_type(mask, T)
+    throw(MOI.UpperBoundAlreadySet{S1,S2}(variable))
+    return
+end
+
+function _throw_if_upper_bound_set(variable, S2, mask, T)
+    upper_mask = mask & _UPPER_BOUND_MASK
+    if iszero(upper_mask)
+        return  # No upper bound set.
+    elseif iszero(_single_variable_flag(S2) & _UPPER_BOUND_MASK)
+        return  # S2 isn't related to the upper bound.
+    end
+    return _throw_if_upper_bound_set_inner(variable, S2, upper_mask, T)
+end
+
+function _lower_bound(
+    set::Union{MOI.GreaterThan,MOI.Interval,MOI.Semicontinuous,MOI.Semiinteger},
+)
+    return set.lower
+end
+
+_lower_bound(set::MOI.EqualTo) = set.value
+
+function _upper_bound(
+    set::Union{MOI.LessThan,MOI.Interval,MOI.Semicontinuous,MOI.Semiinteger},
+)
+    return set.upper
+end
+
+_upper_bound(set::MOI.EqualTo) = set.value
+
+# Use `-Inf` and `Inf` for `AbstractFloat` subtypes.
+_no_lower_bound(::Type{T}) where {T} = zero(T)
+_no_lower_bound(::Type{T}) where {T<:AbstractFloat} = typemin(T)
+_no_upper_bound(::Type{T}) where {T} = zero(T)
+_no_upper_bound(::Type{T}) where {T<:AbstractFloat} = typemax(T)
+
+###
+### SingleVariableConstraints
+###
+### For use in MOI.Utilities.Model
+###
+
+"""
+    struct SingleVariableConstraints{T} <: AbstractVectorBounds
+        variable_indices::Union{Nothing,Set{MOI.VariableIndex}}
+        set_mask::Vector{UInt8}
+        lower::Vector{T}
+        upper::Vector{T}
+    end
+
+A struct for storing SingleVariable-related constraints. Used in `MOI.Model`.
+"""
+mutable struct SingleVariableConstraints{T} <: AbstractVectorBounds
+    variable_indices::Union{Nothing,Set{MOI.VariableIndex}}
+    set_mask::Vector{UInt8}
+    lower::Vector{T}
+    upper::Vector{T}
+end
+
+function SingleVariableConstraints{T}() where {T}
+    return SingleVariableConstraints{T}(nothing, UInt8[], T[], T[])
+end
+
+function Base.:(==)(a::SingleVariableConstraints, b::SingleVariableConstraints)
+    return a.variable_indices == a.variable_indices &&
+           a.set_mask == b.set_mask &&
+           a.lower == b.lower &&
+           a.upper == b.upper
+end
+
+function Base.empty!(b::SingleVariableConstraints)
+    b.variable_indices = nothing
+    empty!(b.set_mask)
+    empty!(b.lower)
+    empty!(b.upper)
+    return b
+end
+
+function Base.resize!(b::SingleVariableConstraints, n)
+    resize!(b.set_mask, n)
+    resize!(b.lower, n)
+    resize!(b.upper, n)
+    return
+end
+
+function MOI.add_variable(b::SingleVariableConstraints{T}) where {T}
+    push!(b.set_mask, 0x00)
+    push!(b.lower, _no_lower_bound(T))
+    push!(b.upper, _no_upper_bound(T))
+    x = MOI.VariableIndex(length(b.set_mask))
+    if b.variable_indices !== nothing
+        push!(b.variable_indices, x)
+    end
+    return x
+end
+
+function MOI.get(b::SingleVariableConstraints, ::MOI.ListOfVariableIndices)
+    if b.variable_indices === nothing
+        return MOI.VariableIndex.(1:length(b.set_mask))
+    end
+    x = collect(b.variable_indices)
```

Comment:
I'm wondering why we don't just put a flag in the mask for deleted variables. What's the point of this `Set` ?